hypothesis_regex.py

```diff
@@ -1,102 +1,268 @@
+            return hs.one_of(*strategies)
+
+    def add_category(self, category):
+        if category == sre.CATEGORY_DIGIT:
+            self._categories |= UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_NOT_DIGIT:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_DIGIT_CATEGORIES
+        elif category == sre.CATEGORY_SPACE:
+            self._categories |= UNICODE_SPACE_CATEGORIES
+            for c in (UNICODE_SPACE_CHARS if self._unicode else SPACE_CHARS):
+                self._whitelist_chars.add(c)
+        elif category == sre.CATEGORY_NOT_SPACE:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_SPACE_CATEGORIES
+            for c in (UNICODE_SPACE_CHARS if self._unicode else SPACE_CHARS):
+                self._blacklist_chars.add(c)
+        elif category == sre.CATEGORY_WORD:
+            self._categories |= UNICODE_WORD_CATEGORIES
+            self._whitelist_chars.add('_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._blacklist_chars.add(c)
+        elif category == sre.CATEGORY_NOT_WORD:
+            self._categories |= UNICODE_CATEGORIES - UNICODE_WORD_CATEGORIES
+            self._blacklist_chars.add('_')
+            if HAS_WEIRD_WORD_CHARS and self._unicode:
+                for c in UNICODE_WEIRD_NONWORD_CHARS:
+                    self._whitelist_chars.add(c)
+
+    def add_chars(self, chars):
+        for c in chars:
+            if self._ignorecase:
+                self._whitelist_chars.add(c.lower())
+                self._whitelist_chars.add(c.upper())
+            else:
+                self._whitelist_chars.add(c)
 
 
 @hs.defines_strategy
 def regex(regex):
-    """Return strategy that generates strings that match given regex."""
-    if hasattr(regex, 'pattern'):
-        regex = regex.pattern
+    """Return strategy that generates strings that match given regex.
+
+    Regex can be either a string or compiled regex (through `re.compile()`).
+
+    You can use regex flags (such as `re.IGNORECASE`, `re.DOTALL` or `re.UNICODE`)
+    to control generation. Flags can be passed either in compiled regex (specify
+    flags in call to `re.compile()`) or inside pattern with (?iLmsux) group.
+
+    Some tricky regular expressions are partly supported or not supported at all.
+    "^" and "$" do not affect generation. Positive lookahead/lookbehind groups
+    are considered normal groups. Negative lookahead/lookbehind groups do not do
+    anything.
+    """
+    if not hasattr(regex, 'pattern'):
+        regex = re.compile(regex)
 
-    codes = sre.parse(regex)
+    pattern = regex.pattern
+    flags = regex.flags
 
-    return ReprWrapperStrategy(_strategy(codes, {}), 'regex(%s)' % repr(regex))
+    codes = sre.parse(pattern)
 
+    return _strategy(codes, Context(flags=flags))
 
-def _strategy(codes, cache):
+
+def _strategy(codes, context):
     if not isinstance(codes, tuple):
```

Comment:
It took me a few cycles through to understand the logic here - it would be nice to clarify this with an explanatory comment, eg 

    # codes is a recursive structure of lists where leaves are (code, value) tuples.
    # We handle the recursion here, and the base case further down.

Then I would also dedent the base-case and add a comment at the top of that.